### DISCUSSION TITLE
[Immunization] How to define the task timing | Content Changes

### DISCUSSION THREAD

**Discussion by dubdabasoduba** [opensrp/fhircore#2142] [Ideas]
# Context

- Immunizations have an ideal timing for when they should be administered. They also have a defined period after which
  the administration is not allowed. This means the immunization is expired.
- Use the `Task.executionPeriod` and `Task.restriction` attributes to define the period of administration and the expiry
  period respectively.

# Implementation Details

### Generate 2 Encounters to record the Immunization recording and Immunization Administration

- We would have 2 Encounters created for each task closed. 1 Encounter will represent the health worker immunization
  record event and the other will represent the immunization administration event.
- Both encounters will be linked to the Task via the `Task.output` attribute.

1. **Encounter 1: Tasks completion event**
    - A Task completion **Encounter** will be used to record the date that the health worker is recording the
      Immunization.
        - The `Encounter.period.start` and `Encounter.period.end` attributes will hold the current date that the data is
          recorded. i.e. the date will be `now()`
        - The `Encounter.type` will hold the vaccination code which
          is `{"coding":[{"system":"http://snomed.info/sct","code":"41000179103","display":"Immunization record (record artifact)"}]}`
        - Link the **Encounter** to the specific task via the `Task.output` attribute. This link will be needed during
          Search, in order to get the Immunization administration date from the **Immunization**, tied to the *
          *Encounter**
            - The `Task.output.type` attribute is set
              to `{"coding":[{"system":"http://snomed.info/sct","code":"41000179103","display":"Immunization record (record artifact)"}]}`
            - The `Task.output.value` attribute is set to the **Encounter** reference
              e.g. `{"reference":"Encounter/<string-uuid>"}`
            - Here is a complete example of the `Task.output`
              attribute `{"type":{"coding":[{"system":"http://snomed.info/sct","code":"41000179103","display":"Immunization record (record artifact)"}]},"value":{"reference":"Encounter/<string-uuid>"}}`

2. **Encounter 2: Immunization administration event**
    - A Immunization administration **Encounter** will be used to record the date the immunization was administered.
      This will mostly be found on the patient's vaccination cards.
        - The `Encounter.period.end` attributes will hold the date the immunization was administered.
        - The `Encounter.type` will hold the vaccination code which
          is `{"coding":[{"system":"http://snomed.info/sct","code":"33879002","display":"Administration of vaccine to produce active immunity (procedure)"}]}`
        - The `Encounter.reasonCode` will hold the code to represent the Specific Task (Immunization). This should be
          found on the `Tasks.reasonCode` attribute.
        - Link the **Encounter** to the specific **Immunization record encounter** a.k.a **Encounter 1** via the `Encounter.partOf` attribute. This link will be needed during
          Search, in order to get the Immunization administration date from the **Immunization**, tied to the *
          *Encounter**
            - Get the `Task.output.type` attribute is set
              to `{"coding":[{"system":"http://snomed.info/sct","code":"41000179103","display":"Immunization record (record artifact)"}]}`
            - Using the `Encounter.partOf` attribute find the encounter that references the Id from on the `Task.output`attribute. 
            - Find the Immunization resource that references the **Encounter 2 (Administration Encounter)**

### Generating an Immunization resource to track the immunization given

- We would generate an Immunization resource instead of the _Immunization record Observation_
- The Immunization will have the following attributes
    - The `Immunization.status` -- This is the status of the immunization resources. It will always be marked as
      completed.
    - The `Immunization.encounter` -- This is a reference to the Tasks completion Encounter which is the Immunization Administration Encounter.
    - The `Immunization.occurrenceDatetime` -- This is the date the immunization was administered. 
    - The `Immunization.reasonCode` -- Holds the code to represent the Specific Task (Immunization). This should be
          found on the `Tasks.reasonCode` attribute.

## How to use the `Task.restriction.period`

- The `Task.restriction.period.end` attribute is used to define when the task expires and cannot be performed anymore.
  For e.g. most of the immunization data might be irrelevant if recorded after 5 yrs _(for child immunizations)_.
    - The `Task.restriction.period.end` attribute will be set to the date the child gets to 5 years. This date can also
      be used to restrict the max date when recording backdated information
    - The expiry implementation has been implemented here https://github.com/opensrp/fhircore/issues/1703
- The `Task.restriction.period.start` attribute should always be the same as the date the Task was initially supposed to
  be DUE/start
    - In most tasks without prerequisites this date will be the same as the `Task.excecutionPeriod.start`

## Required decisions
- Review the Proposed option versus the different schools of thought before the implementation
  - Document the decision in a comment on this discussion

**Comment by pld** [comment]
For the encounter/encounter/immunization strategy, what if you give the immunization at the same time you record it?

**Reply by pld** [reply]
Chatted w/@dubdabasoduba, you'd want to still store 2 encounters so you can write indicators independently. I think the encounter/encounter/immunization strategy makes sense

**Comment by dubdabasoduba** [comment]
# Other options not considered

## Immunization record Encounter

- A Task completion **Encounter** will be used to record the date that the health worker is recording the Immunization.
    - The `Encounter.period.start` and `Encounter.period.end` attributes will hold the current date that the data is
      recorded. i.e. the date will be `now()`
    - The `Encounter.type` will hold the vaccination code which
      is `{"coding":[{"system":"http://snomed.info/sct","code":"41000179103","display":"Immunization record (record artifact)"}]}`
    - Link the **Encounter** to the specific task via the `Task.output` attribute. This link will be needed during
      Search, in order to get the Immunization administration date from the **Observations**, tied to the **Encounter**
        - The `Task.output.type` attribute is set
          to `{"coding":[{"system":"http://snomed.info/sct","code":"41000179103","display":"Immunization record (record artifact)"}]}`
        - The `Task.output.value` attribute is set to the **Encounter** reference
          e.g. `{"reference":"Encounter/<string-uuid>"}`
        - Here is a complete example of the `Task.output`
          attribute `{"type":{"coding":[{"system":"http://snomed.info/sct","code":"41000179103","display":"Immunization record (record artifact)"}]},"value":{"reference":"Encounter/<string-uuid>"}}`

## Immunization administration Observation

- Create an administration **Observation** for Immunization if the recording is done through the **Record all** button
  on the profile cards.
    - if the recording is done on the individual tasks form then we only generate one Observation.
    - The Observation will have the following attributes
        - `category` -- This is a static code to mark the observation as a vaccination
          Observation `{"coding":[{"system":"http://snomed.info/sct","code":"33879002","display":"Administration of vaccine to produce active immunity (procedure)"}]}`
        - `code` -- This is a code to represent the Specific Task (Immunization). This should be found on
          the `Tasks.reasonCode` attribute
        - `effectiveDateTime` -- Represents the immunization administration date.
        - `encounter` -- A reference to the Task administration **Encounter**
    - Here is an Observation
      example `{"resourceType":"Observation","id":"<uuid>","status":"final","category":[{"coding":[{"system":"http://snomed.info/sct","code":"33879002","display":"Administration of vaccine to produce active immunity (procedure)"}]}],"code":{"coding":[{"system":"http://snomed.info/sct","code":"42284007","display":"BCG vaccination"}],"text":"BCG vaccination"},"effectiveDateTime":"<administration-date>","encounter":{"reference":"Encounter/<encounter-id>"}}`

**Comment by dubdabasoduba** [comment]
# Immunization Record Encounter Sample
```json
  {
    "resourceType": "Encounter",
    "id": "14e2ae52-32fc-4507-8736-1177cdaafe90",
    "identifier": {
      "use": "official",
      "value": "6cf76a14-d701-4054-ae1f-4eb387f08e8f"
    },
    "status": "finished",
    "type": [
      {
        "coding": [
          {
            "system": "http://snomed.info/sct",
            "code": "41000179103",
            "display": "Immunization record (record artifact)"
          }
        ]
      }
    ],
    "subject": {
      "reference": "Patient/3e3d698a-4edb-48f9-9330-2f1adc0635d1"
    },
    "period": {
      "start": "2021-10-15T00:00:00+00:00",
      "end": "2021-10-15T00:00:00+00:00"
    }
  }
```

**Comment by dubdabasoduba** [comment]
# Immunization Administration Encounter sample
```json
  {
    "resourceType": "Encounter",
    "id": "ebcfb052-47ed-4846-bb13-0f136cdc53a1",
    "identifier": {
      "use": "official",
      "value": "4b62fff3-6010-4674-84a2-71f2bbdbf2e5"
    },
    "status": "finished",
    "type": [
      {
        "coding": [
          {
            "system": "http://snomed.info/sct",
            "code": "33879002",
            "display": "Administration of vaccine to produce active immunity (procedure)"
          }
        ]
      }
    ],
    "subject": {
      "reference": "Patient/3e3d698a-4edb-48f9-9330-2f1adc0635d1"
    },
    "period": {
      "end": "2021-10-01T00:00:00+00:00"
    },
    "partOf": {
      "reference": "Encounter/14e2ae52-32fc-4507-8736-1177cdaafe90"
    }
  }
```

**Comment by dubdabasoduba** [comment]
# Immunization Resource sample
```json
  {
    "resourceType": "Immunization",
    "id": "41921cfe-5074-4eec-925f-2bc581237660",
    "identifier": {
      "use": "official",
      "value": "6a637a79-df7b-4cc9-93b2-f73f965c31ab"
    },
    "status": "completed",
    "patient": {
      "reference": "Patient/3e3d698a-4edb-48f9-9330-2f1adc0635d1"
    },
    "encounter": {
      "reference": "Encounter/14e2ae52-32fc-4507-8736-1177cdaafe90"
    },
    "occurrenceDateTime": "2021-10-01T00:00:00+00:00"
  }
```

**Comment by dubdabasoduba** [comment]
# OPV 0 resource sample
```json
  {
    "resourceType": "Task",
    "id": "650203d2-f327-4eb4-a9fd-741e0ce29c3f",
    "identifier": [
      {
        "use": "official",
        "value": "ad17cda3-0ac8-43c5-8d9a-9f3adee45e2b"
      }
    ],
    "basedOn": [
      {
        "reference": "CarePlan/28d7542c-ba08-4f16-b6a2-19e8b5d4c229"
      }
    ],
    "status": "requested",
    "intent": "plan",
    "priority": "routine",
    "code": {
      "coding": [
        {
          "system": "http://snomed.info/sct",
          "code": "33879002",
          "display": "Administration of vaccine to produce active immunity (procedure)"
        }
      ]
    },
    "description": "OPV 0 at 0 d vaccine",
    "for": {
      "reference": "Patient/3e3d698a-4edb-48f9-9330-2f1adc0635d1"
    },
    "executionPeriod": {
      "start": "2021-10-01T00:00:00+00:00",
      "end": "2021-10-15T00:00:00+00:00"
    },
    "authoredOn": "2023-03-28T10:46:59+00:00",
    "requester": {
      "reference": "Practitioner/3812"
    },
    "owner": {
      "reference": "Practitioner/3812"
    },
    "reasonCode": {
      "coding": [
        {
          "system": "http://snomed.info/sct",
          "code": "111164008",
          "display": "Poliovirus vaccine"
        }
      ],
      "text": "OPV"
    },
    "reasonReference": {
      "reference": "Questionnaire/9b1aa23b-577c-4fb2-84e3-591e6facaf82"
    },
    "output": [
      {
        "type": {
          "coding": [
            {
              "system": "http://snomed.info/sct",
              "code": "41000179103",
              "display": "Immunization record (record artifact)"
            }
          ]
        },
        "value": {
          "reference": "Encounter/14e2ae52-32fc-4507-8736-1177cdaafe90"
        }
      }
    ]
  }
```

**Comment by dubdabasoduba** [comment]
# OPV 1 resource sample
```json
  {
    "resourceType": "Task",
    "id": "a9100c01-c84b-404f-9d24-9b830463a152",
    "identifier": [
      {
        "use": "official",
        "value": "a20e88b4-4beb-4b31-86cd-572e1445e5f3"
      }
    ],
    "basedOn": [
      {
        "reference": "CarePlan/28d7542c-ba08-4f16-b6a2-19e8b5d4c229"
      }
    ],
    "partOf": {
      "reference": "Task/650203d2-f327-4eb4-a9fd-741e0ce29c3f"
    },
    "status": "requested",
    "intent": "plan",
    "priority": "routine",
    "code": {
      "coding": [
        {
          "system": "http://snomed.info/sct",
          "code": "33879002",
          "display": "Administration of vaccine to produce active immunity (procedure)"
        }
      ]
    },
    "description": "OPV 1 at 6 wk vaccine",
    "for": {
      "reference": "Patient/3e3d698a-4edb-48f9-9330-2f1adc0635d1"
    },
    "executionPeriod": {
      "start": "2021-11-12T00:00:00+00:00",
      "end": "2026-11-11T00:00:00+00:00"
    },
    "authoredOn": "2023-03-28T10:46:59+00:00",
    "requester": {
      "reference": "Practitioner/3812"
    },
    "owner": {
      "reference": "Practitioner/3812"
    },
    "reasonCode": {
      "coding": [
        {
          "system": "http://snomed.info/sct",
          "code": "111164008",
          "display": "Poliovirus vaccine"
        }
      ],
      "text": "OPV"
    },
    "reasonReference": {
      "reference": "Questionnaire/9b1aa23b-577c-4fb2-84e3-591e6facaf82"
    },
    "input": [
      {
        "type": {
          "coding": [
            {
              "system": "http://snomed.info/sct",
              "code": "900000000000457003",
              "display": "Reference set attribute (foundation metadata concept)"
            }
          ]
        },
        "value": {
          "reference": "Task/650203d2-f327-4eb4-a9fd-741e0ce29c3f"
        }
      },
      {
        "type": {
          "coding": [
            {
              "system": "http://snomed.info/sct",
              "code": "371154000",
              "display": "Dependent (qualifier value)"
            }
          ]
        },
        "value": 28
      }
    ],
    "output": [
      {
        "type": {
          "coding": [
            {
              "system": "http://snomed.info/sct",
              "code": "41000179103",
              "display": "Immunization record (record artifact)"
            }
          ]
        },
        "value": {
          "reference": "Encounter/14e2ae52-32fc-4507-8736-1177cdaafe90"
        }
      }
    ]
  }
```
